R/makeTreeSummarizedExperimentFromBiom.R

```diff
@@ -224,3 +244,77 @@ makeTreeSummarizedExperimentFromBiom <- function(obj, ...){
     }
     return(colname)    
 }
+
+# Detect and clean non wanted characters from Taxonomy data if needed.
+.detect_taxa_artifacts_and_clean <- function(x, patterns) {
+    
+    # No cleaning
+    if (is.null(patterns)) {
+        warning("No Taxonomy cleaning was done to the data.",
+                call. = FALSE)
+    }
+    # Automatic cleaning
+    else if (patterns=="auto") {
+        # General regex pattern
+        PATTERN <- "[[:alnum:]]|-|_|\\[|\\]|,|;\\||[[:space:]]"
+        patterns <- .detect_taxa_artifacts(x, PATTERN, invert=TRUE)
+        # No artifacts found
+        if (patterns=="") {
+            warning("No artifacts found in the Taxonomy data.",
+                    call. = FALSE)
+        # Clean from artifacts
+        } else {
+            x <- .clean_from_artifacts(x, patterns)
+            warning("The following artifacts: '", patterns, "' were cleaned from 
+                    the Taxonomy data.",
+                    call. = FALSE)
+        }
+    # Clean with the character or regex provided
+    } else {
+        pattern <- .detect_taxa_artifacts(x, patterns=patterns)
+        # patterns provided not found
+        if (pattern=="") {
+            warning("The '", patterns, "' provided at 'cleanTaxaPattern' were
+                    not found in the Taxonomy data.",
+                    call. = FALSE)
+        # patterns found and cleaned
+        } else {
+            x <- .clean_from_artifacts(x, pattern)
+            # warn what was cleaned
+            warning("The following artifacts: '", pattern, "' were cleaned from 
+                    the Taxonomy data.",
+                    call. = FALSE)
+        }
+    }
+    return(x)
+}
+
+# Helper function for detecting taxa artifacts 
+.detect_taxa_artifacts <- function(
+        x,
+        patterns,
+        invert=FALSE) {
+    if (is(x, "list")) {
+        patterns <- lapply(x, function(x_sub) {
+            grep(patterns,
+                 x_sub[[1]] %>% stringr::str_split("") %>% unlist(),
+                 invert = invert, value = TRUE) %>% unique()
+        }) %>% unlist() %>% unique() %>% paste0(collapse = "")
+    } else if (is(x, "data.frame")){
+        patterns <- apply(x, 2, function(x_sub) {
+            grep(patterns,
+                 x_sub %>% stringr::str_split("") %>% unlist(),
+                 invert = invert, value = TRUE) %>% unique()
+        }) %>% unlist() %>% unique() %>% paste0(collapse = "")
+    }
+    return(patterns)
+}
+
+.clean_from_artifacts <- function(x, patterns) {
+    if (is(x, "list")) {
+        x <- lapply(x, gsub, pattern = patterns, replacement = "")
+    } else if (is(x, "data.frame")) {
+        x <- apply(x, 2, gsub, pattern = patterns, replacement = "")
+    }
```

Comment:
also add "else" statement to ensure that all cases are covered. Even if it does not perform specific  operations.

Or if the type must be list or data.frame then this could throw a warning about a wrong type (although the code should catch that already earlier in principle)

Reply:
What I understood from original implementation is that `feature_data` could be a list, dataframe or null. I didn't choose to call any cleaning when it's NULL, `feature_data`  is constructed as `S4Vectors::make_zero_col_DFrame(nrow(counts))` which means empty content with nothing to clean?
And I guess the code does catch that earlier [at](https://github.com/microbiome/mia/blob/issue_303/R/makeTreeSummarizedExperimentFromBiom.R#L123) ?

Reply:
ok?